shutter/low-level.md

```diff
@@ -16,19 +16,248 @@ This section describes the functionality of the applications involved in the pro
 
 ### Keyper
 
-> :construction: :construction: :construction:
->
-> Todo
->
-> :construction: :construction: :construction:
+Each keyper has a private key `keyper_private_key` to an Ethereum address `keyper_address`. During a setup phase, they have generated `eon_secret_key_share` and a sequence of `eon_public_key_shares`.
+
+#### Chain Monitoring
+
+At all times, the keyper monitors the chain. They keep track of the current slot `s`. They also watch the `keyperSetManager = IKeyperSetManager(KEYPER_SET_MANAGER_ADDRESS)` for
+
+- the current eon `eon = keyperSetManager.getKeyperSetIndexBySlot(s)`,
+- the active keyper set contract `keyperSetContract = IKeyperSet(keyperSetManager.getKeyperSetAddress(eon))`,
+- the active keyper set `keypers = keyperSetContract.getMembers()`, and
+- the threshold `threshold = keyperSetContract.getThreshold()`.
+
+Lastly, they monitor the `validatorRegistry = IValidatorRegistry(VALIDATOR_REGISTRY_ADDRESS)` for the set of indices of the participating validators `participating_validator_indices = get_participating_validators(state)` where `state` refers to the chain state.
+
+#### Slot Processing
+
+At the beginning of each slot `s`, the keyper checks if `keyper_address` is an element of `keypers`. If not, they suspend slot processing until the start of the next slot.
+
+Otherwise, they check if the block proposer of slot `s` is registered in the Validator Registry, i.e., if their validator index is an element of `participating_validator_indices`. If they are not, they suspend slot processing until the start of the next slot.
+
+Otherwise, they fetch the transactions `txs = get_next_transactions(state, eon, tx_pointer)` where `tx_pointer` is a local variable. `tx_pointer` is `0` for the start slot of eon `eon` as defined in the [Keyper Set Manager section](#keyper-set-manager). `tx_pointer` is updated as described in the [Decryption Keys Processing section](#decryption-keys-processing).
```

Comment:
```suggestion
Otherwise, they fetch the transactions `txs = get_next_transactions(state, eon, tx_pointer)` where `tx_pointer` is a local variable. `tx_pointer` is `0` for the start slot of `eon` as defined in the [Keyper Set Manager section](#keyper-set-manager). `tx_pointer` is updated as described in the [Decryption Keys Processing section](#decryption-keys-processing).
```